core/src/main/java/org/kiwix/kiwixmobile/core/page/bookmark/BookmarksActivity.kt

```diff
@@ -0,0 +1,159 @@
+package org.kiwix.kiwixmobile.core.page.bookmark
+
+import android.os.Bundle
+import android.view.Menu
+import android.view.MenuItem
+import android.view.View
+import androidx.appcompat.view.ActionMode
+import androidx.appcompat.view.ActionMode.Callback
+import androidx.appcompat.widget.SearchView
+import androidx.lifecycle.Observer
+import androidx.lifecycle.ViewModelProvider
+import androidx.recyclerview.widget.LinearLayoutManager
+import androidx.recyclerview.widget.RecyclerView
+import io.reactivex.disposables.CompositeDisposable
+import kotlinx.android.synthetic.main.activity_bookmarks.bookmarks_switch
+import kotlinx.android.synthetic.main.activity_bookmarks.no_bookmarks
+import kotlinx.android.synthetic.main.activity_history.recycler_view
+import kotlinx.android.synthetic.main.layout_toolbar.toolbar
+import org.kiwix.kiwixmobile.core.R
+import org.kiwix.kiwixmobile.core.base.BaseActivity
+import org.kiwix.kiwixmobile.core.di.components.CoreComponent
+import org.kiwix.kiwixmobile.core.extensions.ActivityExtensions.coreActivityComponent
+import org.kiwix.kiwixmobile.core.extensions.ActivityExtensions.viewModel
+import org.kiwix.kiwixmobile.core.page.bookmark.adapter.BookmarkDelegate.BookmarkItemDelegate
+import org.kiwix.kiwixmobile.core.page.bookmark.adapter.BookmarkItem
+import org.kiwix.kiwixmobile.core.page.bookmark.adapter.BookmarksAdapter
+import org.kiwix.kiwixmobile.core.page.bookmark.adapter.BookmarksAdapter.OnItemClickListener
+import org.kiwix.kiwixmobile.core.page.bookmark.viewmodel.BookmarkState
+import org.kiwix.kiwixmobile.core.page.bookmark.viewmodel.BookmarkViewModel
+import org.kiwix.kiwixmobile.core.page.viewmodel.Action.Exit
+import org.kiwix.kiwixmobile.core.page.viewmodel.Action.ExitActionModeMenu
+import org.kiwix.kiwixmobile.core.page.viewmodel.Action.Filter
+import org.kiwix.kiwixmobile.core.page.viewmodel.Action.OnItemClick
+import org.kiwix.kiwixmobile.core.page.viewmodel.Action.OnItemLongClick
+import org.kiwix.kiwixmobile.core.page.viewmodel.Action.UserClickedDeleteButton
+import org.kiwix.kiwixmobile.core.page.viewmodel.Action.UserClickedDeleteSelectedPages
+import org.kiwix.kiwixmobile.core.page.viewmodel.Action.UserClickedShowAllToggle
+import org.kiwix.kiwixmobile.core.utils.SimpleTextListener
+import javax.inject.Inject
+
+class BookmarksActivity : OnItemClickListener, BaseActivity() {
```

Comment:
This activity duplicates too much of what is in HistoryActivity.

Have you much experience with generics? Essentially I want to see an abstract base class

```
<T:Page>PageActivity<T>
```

and implementations
```
HistoryActivity: PageActivity<HistoryItem>{ }
BookmarkActivity: PageActivity<BookmarkItem>{ }
```
and in the implementations they will only contain the parts that absolutely have to differ.

A similar methodology may be applied to ViewModel and State.